src/VisualStudio/Core/Impl/SolutionExplorer/AnalyzerItem/AnalyzerItemSource.cs

```diff
@@ -2,237 +2,171 @@
-                    _analyzersFolder.Workspace.WorkspaceChanged -= Workspace_WorkspaceChanged;
-                    break;
-
-                case WorkspaceChangeKind.ProjectAdded:
-                case WorkspaceChangeKind.ProjectReloaded:
-                case WorkspaceChangeKind.ProjectChanged:
-                    if (e.ProjectId == _analyzersFolder.ProjectId)
-                    {
-                        UpdateAnalyzers();
-                    }
-
-                    break;
-
-                case WorkspaceChangeKind.ProjectRemoved:
-                    if (e.ProjectId == _analyzersFolder.ProjectId)
-                    {
-                        _analyzersFolder.Workspace.WorkspaceChanged -= Workspace_WorkspaceChanged;
-                    }
-
-                    break;
-            }
-        }
-
-        private void UpdateAnalyzers()
-        {
-            if (_analyzerItems == null)
-            {
-                // The set of AnalyzerItems hasn't been realized yet. Just signal that HasItems
-                // may have changed.
+internal sealed class AnalyzerItemSource : IAttachedCollectionSource
+{
+    private readonly AnalyzersFolderItem _analyzersFolder;
+    private readonly IAnalyzersCommandHandler _commandHandler;
 
-                NotifyPropertyChanged(nameof(HasItems));
-                return;
-            }
+    private readonly BulkObservableCollection<AnalyzerItem> _items = [];
 
-            var project = _analyzersFolder.Workspace
-                            .CurrentSolution
-                            .GetProject(_analyzersFolder.ProjectId);
+    private readonly CancellationTokenSource _cancellationTokenSource = new();
+    private readonly AsyncBatchingWorkQueue _workQueue;
 
-            if (project != null &&
-                project.AnalyzerReferences != _analyzerReferences)
-            {
-                _analyzerReferences = project.AnalyzerReferences;
+    private IReadOnlyCollection<AnalyzerReference>? _analyzerReferences;
 
-                _analyzerItems.BeginBulkOperation();
+    private Workspace Workspace => _analyzersFolder.Workspace;
+    private ProjectId ProjectId => _analyzersFolder.ProjectId;
 
-                var itemsToRemove = _analyzerItems
-                                        .Where(item => !_analyzerReferences.Contains(item.AnalyzerReference))
-                                        .ToArray();
+    public AnalyzerItemSource(
+        AnalyzersFolderItem analyzersFolder,
+        IAnalyzersCommandHandler commandHandler,
+        IAsynchronousOperationListenerProvider listenerProvider)
+    {
+        _analyzersFolder = analyzersFolder;
+        _commandHandler = commandHandler;
 
-                var referencesToAdd = GetFilteredAnalyzers(_analyzerReferences, project)
-                                        .Where(r => !_analyzerItems.Any(item => item.AnalyzerReference == r))
-                                        .ToArray();
+        _workQueue = new AsyncBatchingWorkQueue(
```

Comment:
>AsyncBatchingWorkQueue [](http://example.com/codeflow?start=25&length=22)

ABWQ to the rescue again!!! Will there be a user-facing change here due to the delay between the Workspace event and the collection update?

Reply:
Technically yes. But so minor is unnoticible